### PR TITLE
feat: initial api/v2 support

### PR DESCRIPTION
**Pull request by p-shahi**
- [X] bump ic-ref to `0.16.0`
- [X] update tests
- [x] update docs
- [x] update icx
- [x] update doc comments

### CODE REVIEW

File: ic-agent/src/agent/mod.rs [hunk 1]
```diff
@@ -52,6 +53,12 @@ const IC_STATE_ROOT_DOMAIN_SEPARATOR: &[u8; 14] = b"\x0Dic-state-root";
 ///
 /// Any error returned by these methods will bubble up to the code that called the [Agent].
 pub trait ReplicaV1Transport {
```
Comment (hansl) [inline ic-agent/src/agent/mod.rs:55]: Create a new `V2` version of this.

Comment (p-shahi) [reply ic-agent/src/agent/mod.rs:55]: Will do

File: ic-agent/src/agent/mod.rs [hunk 2]
```diff
@@ -250,7 +269,71 @@ impl Agent {
...
+        let bytes = self
+            .transport
+            .query(effective_canister_id, serialized_bytes)
+            .await?;
+        serde_cbor::from_slice(&bytes).map_err(AgentError::InvalidCborData)
+    }
+
+    async fn read_state_endpoint<A>(
+        &self,
+        effective_canister_id: Principal,
```
Comment (hansl) [inline ic-agent/src/agent/mod.rs:304]: You don't always have a canister id though :(

Comment (nomeata) [reply ic-agent/src/agent/mod.rs:304]: (Just copying here what I said on slack)

I think it’s fine to expose this functionality only when an effective canister id is provided. After all, you need to indicate where you want to read the state! And current use cases for `read_state` (call status, controller, module_hash) are related to a specific canister.

File: icx/src/main.rs [hunk 3]
```diff
@@ -405,26 +429,27 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
-                    line = input.pop_front().unwrap(); // request id.
-                    let request_id = RequestId::from_str(&line)?;
-                    input.pop_front().unwrap(); // empty line.
-                    line = input.pop_front().unwrap(); // envelope
-                    let envelope = hex::decode(line)?;
-                    transport.submit(envelope, request_id).await?;
-                    eprint!("Request ID: ");
-                    println!("0x{}", hex::encode(request_id.as_slice()));
-                }
+                // "read" => {
```
Comment (hansl) [inline icx/src/main.rs:434]: ?

Comment (hansl) [reply icx/src/main.rs:434]: This code is used, we need to update it as well.

Comment (p-shahi) [reply icx/src/main.rs:434]: I forgot to update this.

File: ic-agent/src/agent/mod.rs [hunk 4]
```diff
@@ -122,8 +128,8 @@ pub trait ReplicaV1Transport {
 ///
 ///   // Create a call to the management canister to create a new canister ID,
 ///   // and wait for a result.
-///   let response = agent.update(&management_canister_id, "create_canister")
-///     .with_arg(&Encode!()?)  // Empty Candid.
+///   let response = agent.update(&management_canister_id, "provisional_create_canister_with_cycles")
+///     .with_arg(&Encode!(&Argument { amount: None})?)
```
Comment (hansl) [inline ic-agent/src/agent/mod.rs:132]: Nit: space before the `}`

**Review by hansl — APPROVED** [review]

**Pull request merged**